aleph-client/src/lib.rs

```diff
@@ -105,21 +109,40 @@ pub fn create_custom_connection<Client: FromStr + RpcClient>(
     }
 }
 
-pub fn send_xt(connection: &Connection, xt: String, xt_name: &'static str, tx_status: XtStatus) {
-    let block_hash = connection
-        .send_extrinsic(xt, tx_status)
-        .expect("Could not send extrinsic")
-        .expect("Could not get tx hash");
-    let block_number = connection
-        .get_header::<Header>(Some(block_hash))
-        .expect("Could not fetch header")
-        .expect("Block exists; qed")
-        .number;
-    info!(
-        target: "aleph-client",
-        "Transaction {} was included in block {}.",
-        xt_name, block_number
-    );
+/// `panic`able utility wrapper for `try_send_xt`.
+pub fn send_xt<T: Encode>(
+    connection: &Connection,
+    xt: UncheckedExtrinsicV4<T>,
+    xt_name: Option<&'static str>,
+    xt_status: XtStatus,
+) -> Option<H256> {
+    try_send_xt(connection, xt, xt_name, xt_status).expect("Should manage to send extrinsic")
+}
+
+/// Sends transaction `xt` using `connection`. If `tx_status` is either `Finalized` or `InBlock`
+/// additionally returns hash of the containing block. `xt_name` is used only for logging purposes.
+/// Recoverable.
+pub fn try_send_xt<T: Encode>(
+    connection: &Connection,
+    xt: UncheckedExtrinsicV4<T>,
+    xt_name: Option<&'static str>,
+    xt_status: XtStatus,
+) -> ApiResult<Option<H256>> {
+    let hash = connection
+        .send_extrinsic(xt.hex_encode(), xt_status)?
+        .ok_or_else(|| Error::Other(String::from("Could not get tx/block hash").into()))?;
+
+    match xt_status {
+        XtStatus::Finalized | XtStatus::InBlock => {
+            info!(target: "aleph-client",
+                "Transaction `{}` was included in block with hash {}.",
+                xt_name.unwrap_or_default(), hash);
+            Ok(Some(hash))
+        }
+        // Other variants either do not return (see https://github.com/scs/substrate-api-client/issues/175)
+        // or return xt hash, which is kinda useless here.
+        _ => Ok(None),
```

Comment:
Why is it useless? Even if user didn't specify that he wants to wait for specific xt_status I think it's still useful to return the block hash the txn was included in. Why throw away that information?

Reply:
Hash of the extrinsic is the thing that you already had before calling this method. Note, that only `InBlock` and `Finalize` variants provide block hash.

Reply:
and what is funnier, there is no documentation in `substrate-api-client` on it

and what is even funnier, in different examples in their repository there are different interpretations xD

but I checked and it's indeed block hash

Reply:
so, if it is a block hash, why not return it in all cases?

Reply:
no no, it is a block hash only for `InBlock` and `Finalize`; for other it is extrinsic hash